lib/kernel.js

```diff
@@ -16,34 +16,89 @@ import store from "./store";
 import WatchesStore from "./store/watches";
 import OutputStore from "./store/output";
 import HydrogenKernel from "./plugin-api/hydrogen-kernel";
+import InputView from "./input-view";
+import KernelTransport from "./kernel-transport";
+import ZMQKernel from "./zmq-kernel";
+import type { ResultsCallback } from "./kernel-transport";
+
+interface KernelMiddlewareStack {
+  interrupt: () => void;
+  shutdown: () => void;
+  restart: (onRestarted: ?Function) => void;
+  execute: (
+    code: string,
+    callWatches: boolean,
+    onResults: ResultsCallback
+  ) => void;
+  complete: (code: string, onResults: ResultsCallback) => void;
+  inspect: (
+    code: string,
+    cursorPos: number,
+    onResults: ResultsCallback
+  ) => void;
+}
 
 export default class Kernel {
-  @observable executionState = "loading";
   @observable inspector = { bundle: {} };
   outputStore = new OutputStore();
 
-  kernelSpec: Kernelspec;
-  grammar: atom$Grammar;
-  language: string;
-  displayName: string;
   watchesStore: WatchesStore;
-  watchCallbacks: Array<Function> = [];
   emitter = new Emitter();
   pluginWrapper: HydrogenKernel | null = null;
+  kernel: ZMQKernel | KernelTransport; // XXX(nikita): need explicit ZMQKernel for plugins to typecheck
+  middlewareStack: KernelMiddlewareStack;
 
-  constructor(kernelSpec: Kernelspec, grammar: atom$Grammar) {
-    this.kernelSpec = kernelSpec;
-    this.grammar = grammar;
-
-    this.language = kernelSpec.language.toLowerCase();
-    this.displayName = kernelSpec.display_name;
+  constructor(kernel: KernelTransport) {
+    this.kernel = kernel;
 
     this.watchesStore = new WatchesStore(this);
+
+    this.middlewareStack = {
```

Comment:
There is no need for this, just move this code into the corresponding methods in Kernel(=HKernel).